### PR TITLE
[query] Add TableMapPartitions node

### PR DESCRIPTION
**Pull request by tpoterba**
To come next:
* Expose in Python, add more tests of Python functionality

### CODE REVIEW

File: hail/src/main/scala/is/hail/expr/ir/Compile.scala [hunk 1]
```diff
@@ -133,3 +134,130 @@ object CompileWithAggregators {
...
+
+    val fb = EmitFunctionBuilder.apply[F](ctx, "stream", argTypeInfo.toFastIndexedSeq, CodeParamType(BooleanInfo))
+    val stepF = fb.apply_method
+    val stepFECB = stepF.ecb
+
+    val er = EmitRegion.default(stepF)
+    val emitter = new Emit(ctx, stepFECB)
+
+    var ir = body
+    ir = LoweringPipeline.compileLowerer(true)(ctx, ir).asInstanceOf[IR].noSharing
```
Comment (patrick-schultz) [inline hail/src/main/scala/is/hail/expr/ir/Compile.scala:186]: Why the var? Can't this just be
```
    val ir = LoweringPipeline.compileLowerer(true)(ctx, body).asInstanceOf[IR].noSharing
```
?

File: hail/src/main/scala/is/hail/expr/ir/Compile.scala [hunk 2]
```diff
@@ -133,3 +134,130 @@ object CompileWithAggregators {
...
+    val didSetup = stepF.genFieldThisRef[Boolean]("didSetup")
+
+    implicit val ecc: EmitStreamContext = EmitStreamContext(stepF)
+    var source: Source[EmitCode] = null
+
+    val pullLabel = CodeLabel()
+    val eosField = stepF.genFieldThisRef[Boolean]("eos")
+    val eosLabel = CodeLabel()
+
+    optStream.addSetup(Code(elementAddress := 0L, eosField := false))
```
Comment (patrick-schultz) [inline hail/src/main/scala/is/hail/expr/ir/Compile.scala:204]: I think it's clearer if you just do these after `source.setup` below.

File: hail/src/main/scala/is/hail/expr/ir/Compile.scala [hunk 3]
```diff
@@ -133,3 +134,130 @@ object CompileWithAggregators {
...
+    ir = LoweringPipeline.compileLowerer(true)(ctx, ir).asInstanceOf[IR].noSharing
+    TypeCheck(ir)
+    InferPType(ir, Env.empty[PType])
+    val returnType = ir.pType.asInstanceOf[PStream].elementType.asInstanceOf[PStruct].setRequired(true)
+
+    val optStream = EmitStream.emit(emitter, ir, stepF, er.region, Env.empty, None);
+
+    val elementAddress = stepF.genFieldThisRef[Long]("elementAddr")
+
+    val didSetup = stepF.genFieldThisRef[Boolean]("didSetup")
```
Comment (patrick-schultz) [inline hail/src/main/scala/is/hail/expr/ir/Compile.scala:195]: This is never initialized. It's probably better not to rely on Java's default initialization behavior. Add `didSetup := false` to the class initializer?

File: hail/src/main/scala/is/hail/expr/ir/Interpretable.scala [hunk 4]
```diff
@@ -1,11 +1,12 @@
 package is.hail.expr.ir
 
-import is.hail.types.virtual.TNDArray
+import is.hail.types.virtual.{TNDArray, TStream}
 
 object Interpretable {
   def apply(ir: IR): Boolean = {
     !ir.typ.isInstanceOf[TNDArray] &&
-      (ir match {
+      !ir.typ.isInstanceOf[TStream] &&
```
Comment (patrick-schultz) [inline hail/src/main/scala/is/hail/expr/ir/Interpretable.scala:8]: Why this? There are interpret rules for most of the stream nodes.

File: hail/src/main/scala/is/hail/expr/ir/Compile.scala (hunk 1 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/Compile.scala:186]: good catch, this is an artifact of now-removed control flow.

File: hail/src/main/scala/is/hail/expr/ir/Interpretable.scala (hunk 4 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/Interpretable.scala:8]: not sure, actually. will revert.

**Comment by cseed** [comment]
Curious, did you benchmark this change, in particular the local/field business?

**Comment by tpoterba** [comment]
no difference: https://gist.github.com/tpoterba/2d3c72db1eba30e601e7a1876a675cd4

**Comment by tpoterba** [comment]
if there's an effect, I expect we'll see it when we do whole-stage codegen, which we can't benchmark now.

**Pull request merged**